### ISSUE TITLE
For loops and conditions vs. apply and logical subsetting

### ISSUE THREAD

**Issue by dmi3kno** [swcarpentry/r-novice-inflammation#250]
I am of strong opinion that introducing `for loops` and `if else` statements early in the teaching program makes more harm to R education then good. I understand that this is done out of desire to maintain consistency between how Python and R are taught, but I would argue that the approach to teaching (and using) the two languages should be different.

I would argue that `for` loops and `if else` statements need to be shelved under the "Advanced" topics towards the end of the teaching lesson and instead a sections on `*apply` family of functions and logical subsetting should be introduced. R is positioning itself as a `vectorized` language (even though under the hood it might be running highly optimized `for` loops in C++), but the R programmer is encouraged to think in terms of vectors and data frames. Introduction of non-vectorized operators breaks that frame and positions R for failure due to apriori lost argument on speed and efficiency. 

Again, if you agree, I volunteer to handpick material on `*apply` operators from another lesson (I will not introduce any new concepts but rather repackages what is already in the very good Software Carpentry curriculum) and rework logical subsetting (and, perhaps, mention vectorized `ifelse()` function) to cater for the need to teach implementation of conditional logic (branching) in R. 

As I said, the loop and branching sections are good, but only as an advanced topic towards the end of the lesson material.

**Comment by RMHogervorst** [comment]
I agree.

**Comment by chendaniely** [comment]
I agree that Python and R are different languages with different domains and should be treated accordingly.

My only concern is if we remove `for` loops until the very end of the lesson, then the entire lesson will need to be restructured.
The current lesson gets users to load data, create a plot, then automatically create multiple plots from multiple datasets.  This requires a `for` loop (unless I am mistaken).

The rationale behind the current lesson order is that `for` loops and conditionals are fundamental in programming (in other languages).  Since this is the `programming with r` lesson and not a `data analysis in r` lesson, the entire lesson is more focused on programming concepts, rather than data analysis concepts.

I am more than happy to continue this discussion, as it does set a foundation for our students.

**Comment by RMHogervorst** [comment]
I guess you could use a apply action in stead of a for loop, but that is just a for loop in hiding. Since that does not really make a difference but would introduce new functions . So that doesn't help in the software carpentry lessons. 
I agree with the programming with r vs data analyses in r rationale.

**Comment by dmi3kno** [comment]
My argument, then, is that there's no such thing as "beginner programmer in R". There's only "beginner analyst in R". It is very rare instance when for-loops need to be written and those shall be reserved to non-rectangular data types. For everything else R has an awesome functional programming toolbox with base::*apply and purrr::map_* families which (although rely on C++ for loops) emphasize the functional aspect of it and hide away the implementation details (which do more harm than good to beginners). This is highly philosophical discussion and I am ready to give in on changing the lesson, if you guys confirm that you taught R with for loops and you tried introducing apply instead and you liked the former better.

**Comment by katrinleinweber** [comment]
Hello! Having myself been enchanted by `purrr::map()` last year, how about a `…-suppl-….Rmd` compares and contrasts it and `base::apply`?

**Comment by katrinleinweber** [comment]
Related to #276, because both [readr & purrr are in the tidyverse](https://www.tidyverse.org/packages/).

**Comment by dmi3kno** [comment]
R is evolving so fast that I no longer want to stand by base::apply(). It should be purrr::map all the way. We tried teaching it in SWC Oslo and it works like a charm. Highly recommend watching Hadley's cupcake rant video: https://www.youtube.com/watch?v=GyNqlOjhPCQ

Also, plenty of resources for teaching purrr, not least by Jenny Bryan

**Comment by CodeRThane** [comment]
I've only been coding for about 6 months now, but here are my 2 cents:
I think for loops (or the apply/map functions) should be taught early in R. My first R project after that involved working with over 100 CSV files. From the data carpentry lesson I had taken, I knew how to work with one CSV at a time, but I had to spend a lot of time on the internet to figure out how to use the apply function before I could make much progress on the project.

I guess what I'm trying to say is that for loops are an important functional tool that programmers need to have at their fingertips, and I think it should be taught early on.

**Comment by HaoZeke** [comment]
The argument made by @dmi3kno makes sense to me as well. @CodeRThane, the idiomatic `R` method is to use `purrr::map()` or `apply` instead of `for` loops as you noticed. I'd be happy to see some concrete PRs for this issue.

**Comment by aforestsomewhere** [comment]
I'm starting to go through old issues - please bear with me as I get to grips with the lesson, and thank you to all contributors to this long-running thread - and wanted to summarise the key points/suggestions and hopefully spark some new discussion/PRs. (I've gone through the previous discussions here, and on related issue [@449](https://github.com/swcarpentry/r-novice-inflammation/issues/449), but feel free to add/correct/comment if things have changed or if other larger-scale edits planned might interact with things here).

It sounds like the consensus arrived at was to revise [03-loops-R.Rmd](https://github.com/swcarpentry/r-novice-inflammation/blob/main/episodes/03-loops-R.Rmd) to replace `for` loops with `purrr::map()` (and possibly also `apply`), also drawing inspiration from the presentation linked above from (@jennybc [Thinking inside the box (45min webinar)](https://www.rstudio.com/resources/webinars/thinking-inside-the-box-you-can-do-that-inside-a-data-frame/)) and the [current supplementary lesson](https://github.com/swcarpentry/r-novice-inflammation/blob/54e86f4969d62aaaad993d0c80bbb7c3ad8a38dd/episodes/15-supp-loops-in-depth.Rmd). My initial thought is that it makes sense to keep the overall goal for episode [03-loops-R.Rmd](https://github.com/swcarpentry/r-novice-inflammation/blob/main/episodes/03-loops-R.Rmd) as being able to apply a function over a list of items, specifically to be able to read in and analyse multiple input files. 

I did note that there was a preference at one point to not adapt this entire lesson over to the Tidyverse ([@276](https://github.com/swcarpentry/r-novice-inflammation/issues/276)), but `purr::map()`  seemed to be favoured overall on this thread? Seeing as `apply()` has been introduced [earlier](https://swcarpentry.github.io/r-novice-inflammation/01-starting-with-data.html), an option might be to clarify the differences between vector/data frame/list structures, possibly outline the benefits of row-wise work, then make an alternate simple example using `purr::map()` instead of the current example of printing vectors (best_practice <- c("Let", "the", "computer", "do", "the", "work")), following by implementing the `analyze` function that learners would have just written, via `purr::map()`

I think a expand/callout on `for` loops could still be useful, as this is instinctive for many people who have coded in other languages. [15-supp-loops-in-depth](https://github.com/swcarpentry/r-novice-inflammation/blob/54e86f4969d62aaaad993d0c80bbb7c3ad8a38dd/episodes/15-supp-loops-in-depth.Rmd) can be merged in here as suggested above, emphasising that `for` loops can exist in R, but that there are good reasons for adopting vectorised approaches from the get-go?
